mindmeld/models/taggers/pytorch_crf.py

```
@@ -0,0 +1,408 @@
+            log_prob.append(score * mask[i].unsqueeze(1) +
+                            log_prob[-1] * (1. - mask[i]).unsqueeze(1))
+
+        if run_backwards:
+            log_prob.reverse()
+
+        return torch.stack(log_prob)
+
+    def compute_marginal_probabilities(self, inputs, mask):
+        # SWITCHING FOR BATCH FIRST DEFAULT
+        dense_W = torch.tile(self.W, dims=(mask.shape[0], 1))
+        out_1 = torch.addmm(self.b, inputs, dense_W)
+        emissions = out_1.reshape((mask.shape[0], -1, self.num_classes))
+        emissions = emissions.transpose(0, 1)
+        mask = mask.transpose(0, 1)
+        alpha = self._compute_log_alpha(emissions, mask, run_backwards=False)
+        beta = self._compute_log_alpha(emissions, mask, run_backwards=True)
+        z = torch.logsumexp(alpha[alpha.size(0) - 1] + self.crf_layer.end_transitions, dim=1)
+        prob = alpha + beta - z.view(1, -1, 1)
+        return torch.exp(prob).transpose(0, 1)
+
+    def set_params(self, **params):
+        self.feat_type = params.get('feat_type', DEFAULT_PYTORCH_CRF_ER_CONFIG['feat_type']).lower()
+        self.feat_num = params.get('feat_num', DEFAULT_PYTORCH_CRF_ER_CONFIG['feat_num'])
+        self.stratify = params.get('stratify_train_val_split',
+                                   DEFAULT_PYTORCH_CRF_ER_CONFIG['stratify_train_val_split'])
+        self.drop_input = params.get('drop_input', DEFAULT_PYTORCH_CRF_ER_CONFIG['drop_input'])
+        self.train_batch_size = params.get('train_batch_size', DEFAULT_PYTORCH_CRF_ER_CONFIG['train_batch_size'])
+        self.patience = params.get('patience', DEFAULT_PYTORCH_CRF_ER_CONFIG['patience'])
+        self.epochs = params.get('epochs', DEFAULT_PYTORCH_CRF_ER_CONFIG['epochs'])
+        self.train_val_split = params.get('train_val_split', DEFAULT_PYTORCH_CRF_ER_CONFIG['train_val_split'])
+        self.optimizer_type = params.get('optimizer_type', DEFAULT_PYTORCH_CRF_ER_CONFIG['optimizer_type']).lower()
```

I think it is better to raise an error if user mentions uppercased string like "Adam", "ADAM" , etc. instead of "adam" as we'll only mention the latter in the docs. Implementation-wise, I think this is not an issue here but if there was a scenario where we propagate the params further down into another method and missed to do lowercasing there; debugging will become difficult.

---

Fixed as mentioned